File: src/sly/speakrecognizer/audio/modules/DctModule.scala
package sly.speakrecognizer.audio.modules

import sly.speakrecognizer.audio.SingleData
import sly.speakrecognizer.audio.Module
import sly.speakrecognizer.audio.Data
import sly.speakrecognizer.audio.MultiData

class DctModule(inputValuesLength: Int = 20, coefficentsLength :Int = 13) extends Module {
	
  val sqrt = Math.sqrt(inputValuesLength)
  val dct = new DCT(inputValuesLength,coefficentsLength - 1);
  	
  override def process(input: Data): Option[Data] = {
    val res = input.iterator.map(d => new SingleData({
        val resAr = Array.ofDim[Double](coefficentsLength);
        val tab = d.asArrayDouble(0)
        dct.transform(tab,resAr,1)
        var sum = 0.0
        tab.foreach{d =>
          sum += d
        }
        resAr(0) = sum / sqrt
        resAr
    })).toArray[Data]
    Some(new MultiData(res))
  }
  
}

File: src/sly/speakrecognizer/modules/soundanalyzer/DctModule.java
package sly.speakrecognizer.modules.soundanalyzer;

import sly.speakrecognizer.math.functions.DCT;
import sly.speakrecognizer.modules.AbstractModule;
import sly.speakrecognizer.modules.Data;
import sly.speakrecognizer.modules.Module;
import sly.speakrecognizer.modules.SoundSamplesData;

public class DctModule extends AbstractModule implements Translator{
	private DCT dct = null;
	private int sizeCoef;
	private double sqrt;
	
	public DctModule(Module output, Parameters parameters) {
		super(output);
	}

	public void putData(Data data) {
		if(output != null){
			output.putData(translate(data));
		}
	}

	public Data translate(Data data) {
		SoundSamplesData sdata = (SoundSamplesData) data;
		double[] tab = sdata.getData(0);
		if(dct == null){
			sizeCoef = 12;
			dct = new DCT(tab.length,sizeCoef);
			sizeCoef += 1;
			sqrt = Math.sqrt(tab.length);
		}else{
			if(dct.getLengthFilters() != tab.length){
				throw new IllegalStateException("Ilość filtrów cech Mel niezgodna z poprzednim ustawieniem");
			}
		}

		double[] res = new double[sizeCoef];
		//DCT
		dct.transform(tab,res,1);
		//Signal Energy
		double sum = 0.0;
		for(double v : tab){
			sum += v;
		}
		res[0] = sum / sqrt;
		return new SoundSamplesData(res,sdata.getFreq());
	}
	
}
